Modules/_struct.c

```diff
@@ -545,12 +545,8 @@ static int
 np_byte(_structmodulestate *state, char *p, PyObject *v, const formatdef *f)
 {
     long x;
-    if (get_long(state, v, &x) < 0)
-        return -1;
-    if (x < -128 || x > 127) {
-        PyErr_SetString(state->StructError,
-                        "byte format requires -128 <= number <= 127");
-        return -1;
+    if (get_long(state, v, &x) < 0 || x < -128 || x > 127) {
```

Comment:
This logic doesn't look right: if `get_long` returns `-1`, we don't necessarily know that that represents an out-of-range error.

For example, on this branch, we have:
```
>>> struct.pack("b", "not a number")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
struct.error: 'b' format requires -128 <= number <= 127
```

On the `main` branch, the error is:
```
struct.error: required argument is not an integer
```

We may need to do some reworking of `get_long` to get a good solution to this issue. With this code, if an overflow error occurs in `PyLong_AsLongAndOverflow` we end up resetting the exception state twice: once in `get_long` and once in `RANGE_ERROR` below, which doesn't seem ideal.

Reply:
Thank you. Both problems are fixed. Now `struct.pack("b", "not a number")` will give the correct error message and the corresponding test cases are updated. The exception state is only set once when an overflow error occurs.

Reply:
Thanks for the update! We're getting there, but this still isn't quite right. In the `RANGE_ERROR` call below, you know that you have an exception that _isn't_ a `StructError`. That unfortunately doesn't tell you much about what the exception is, and where it came from. We want to be using the new exception message only when we're sure that we have an out-of-range error.

For example, on this branch we now have this behaviour:
```python
Python 3.12.0a1+ (heads/gh-98248:856a9bef05, Nov 27 2022, 10:51:22) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> class A:
...     def __index__(self): 1/0
... 
>>> import struct; struct.pack("b", A())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
struct.error: 'b' format requires -128 <= number <= 127
```

Instead, the `ZeroDivisionError` should propagate here, as it does in current Python:
```python
Python 3.10.7 (main, Sep 10 2022, 07:37:51) [Clang 13.0.0 (clang-1300.0.29.30)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> class A:
...     def __index__(self): 1/0
... 
>>> import struct; struct.pack("b", A())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in __index__
ZeroDivisionError: division by zero
```

Reply:
I think it _may_ be enough to replace the `!PyErr_ExceptionMatches(state->StructError)` condition with `PyErr_ExceptionMatches(PyExc_OverflowError)`

Reply:
I think the following contains the right logic (for `np_byte`), and appears to work for me:
```c
static int
np_byte(_structmodulestate *state, char *p, PyObject *v, const formatdef *f)
{
    long x;
    int err = get_long(state, v, &x);

    if (err < 0) {
        if (PyErr_ExceptionMatches(PyExc_OverflowError)) {
            RANGE_ERROR(state, f, 0);
        }
        return -1;
    }
    if (x < -128 || x > 127) {
        RANGE_ERROR(state, f, 0);
    }
    *p = (char)x;
    return 0;
}
```
Other functions would need to be updated accordingly, of course.